README.md
# Floating Player
Floating Player is a Firefox extension to watch videos while you browse the internet

## How to install
https://addons.mozilla.org/pt-BR/firefox/addon/floating-player/


CHANGELOG
3.6 - 2018/04/26
- Changed extension description to make it shorter

3.5 - 2018/04/25
- Changed extension name to "Floating Player" due to removal from Chrome Web Store

3.4 - 2018/04/23
- Removed legacy code
- Set "show related videos" as default option

3.3 - 2018/03/16
- Fixed bug for odd width/height and popup centered

3.2 - 2018/03/06
- Added option "Show chat for YouTube and Twitch live streams"

3.1 - 2018/03/05
- Improved code

3.0 - 2018/02/27
- Improved code

2.0.37 - 2018/02/17
- Fixed bug for Compiz window manager on Ubuntu

2.0.36.1 - 2018/02/10
- Fixed bug for older browsers

2.0.36 - 2018/02/10
- Added secret option to enable the add-on to set borderless/always-on-top popup
  • Download https://github.com/gabrielbarros/floating-player-app/archive/master.zip
  • Unzip it
  • Move the folder "floating-player-app-master" to somewhere safe (Documents, Desktop, etc)
  • Open Google Chrome
  • Go to chrome://extensions/
  • Check "Developer mode"
  • Click the button "Load unpacked extension"
  • Select the folder ...\floating-player-app-master\src (<-- SRC FOLDER!)
  • Go to chrome-extension://ekajjllcmeckibblgckgoceinmmgnfop/options.html
  • Press F9
  • Click "Allow" and test it on YouTube
  • If you want to disable it, press F9 again on options page

2.0.35 - 2018/02/07
- Now ?floating_player=1 is added at the end of the url so you can inject scripts
  and css via tampermonkey

2.0.34 - 2018/01/08
- Added option "Pin temporary tab before opening Helium" (macOS only)
- Added option "Helium app" (macOS only)

2.0.33 - 2018/01/04
- Added option "Keep position and current dimensions of popup when changing video"

2.0.32 - 2017/12/29
- Popup is focused on Chrome

2.0.31 - 2017/12/28
- Port to Firefox

2.0.30 - 2017/12/18
- Added option "Spoof Referrer header to be able to embed certain videos"
- Improved design for instructions, history and options page

2.0.29 - 2017/12/13
- Fixed option "Block mouse events"
- Removed Spanish translation
- Improved instructions page

2.0.28 - 2017/10/11
- Hidden options of ChromeOS and macOS from other OSs

2.0.27 - 2017/10/06
- Improved code

2.0.26 - 2017/10/05
- Fixed translation

2.0.25 - 2017/10/04
- Added option "Keep popup always on top on Chrome OS"

2.0.24 - 2017/10/04
- Added option "Remove borders and title bar on Chrome OS"

2.0.23 - 2017/10/01
- Added option "Block mouse events on YouTube videos"

2.0.22 - 2017/09/30
- Added option "Title for YouTube videos"
- Removed option "Show YouTube playlist counter[...]"

2.0.21 - 2017/09/29
- Fixed new url of Twitch.tv
- Added volume control for Twitch.tv

2.0.20 - 2017/08/31
- Improved instructions page

2.0.19 - 2017/08/28
- Added compatibility with addon for Chrome OS

2.0.18 - 2017/08/19
- Added option "Force full screen"

2.0.17 - 2017/08/18
- Removed menu instructions due to unknown bug

2.0.16 - 2017/08/18
- Added menu instructions

2.0.15 - 2017/08/09
- Improved loop for single videos

2.0.14 - 2017/08/06
- Added support to Pocket

2.0.13 - 2017/08/05
- Fixed incorrect video quality

2.0.12 - 2017/08/05
- Fixed bug for non-embeddable videos

2.0.11 - 2017/08/05
- Focus on popup after updating its url

2.0.10 - 2017/08/04
- Now popup won't resize when pausing/playing
- Fixed popup aspect ratio from 4:3 to non-youtube video

2.0.9 - 2017/08/03
- Code improvements
- Added support to Google search links

2.0.8 - 2017/07/29
- Readded permission to send data to YouTube

2.0.7 - 2017/07/29
- Fixed popup height on Mac OS X
- Code improvements

2.0.6 - 2017/07/28
- Removed needless permissions
- Added support to time params for YouTube videos
- Added autoplay for Facebook
- Added support to Helium on Mac OS X

2.0.5 - 2017/07/27
- Fixed bug when videos play fullscreen

2.0.4 - 2017/07/27
- Removed option "4:3 size"
- 4:3 width is now calculated automatically from 16:9 height

2.0.3 - 2017/07/25
- Design improvements
- Added new recommended resolutions

2.0.2 - 2017/07/22
- Added option "Close current tab when creating a new popup"
- Added support to youtu.be links
- Added new alignment options

2.0.1 - 2017/07/20
- Added option "Animate popup title for YouTube videos"
- Added option "Loop YouTube videos and playlists"
- Added support to Ustream.tv

2.0 - 2017/07/19
- Added option "Enable history"

1.9.10 - 2017/07/18
- Added option "Shuffle YouTube playlist"
- Added option "Playback speed of YouTube videos"
- Added option "Quality of YouTube videos"
- Added option "Volume for YouTube videos"

1.9.9 - 2017/07/13
- Added option "Show fullscreen button on YouTube"
- Added option "Show YouTube logo that lets you open the video in a new tab"
- Added option "Allow control of the YouTube player by keyboard"
- Added autoplay for Vimeo, Smashcast and Vevo

1.9.8 - 2017/07/10
- Added YouTube Gaming
- Added option "Use YouTube TV for all YouTube videos"

1.9.7 - 2017/07/09
- Added option "Show YouTube playlist counter in popup title"
- Added button "Source code"
- Added support to YouTube channel and search
- Added option "Show video player controls on YouTube"
- Added option "Show video title on YouTube"
- Added option "Enable privacy-enhanced mode using youtube-nocookie.com"
- Removed permission to read the browsing history

1.9.6 - 2017/07/07
- Floating Player is open source now

1.9.5 - 2017/06/28
- Added option "Keep current popup instead of opening a new one"


src/js/options.js
// Translation strings
var strOptions = getText('options');
setHtml($$('title'), strOptions);
setHtml($$('h1 a'), strOptions);
setHtml($$('label[for="align"]'), '@align');

var $option = $$$('option');
setHtml($option[0], '@top_left');
setHtml($option[1], '@top_right');
setHtml($option[2], '@bottom_left');
setHtml($option[3], '@bottom_right');
setHtml($option[4], '@top_center');
setHtml($option[5], '@bottom_center');
setHtml($option[6], '@left_center');
setHtml($option[7], '@right_center');
setHtml($option[8], '@center');

setHtml($$('label[for="width"]'), '@size');
setHtml($('recommended'), '@recommended');
setHtml($$('label[for="horizontal-margin"]'), '@horizontal_margin');
setHtml($$('label[for="vertical-margin"]'), '@vertical_margin');

setHtml($$('label[for="color"]'), '@color');
setHtml($option[9], '@red');
setHtml($option[10], '@white');

setHtml($$('label[for="speed"]'), '@speed');
setHtml($option[14], '@normal');

setHtml($$('label[for="quality"]'), '@quality');
setHtml($$('label[for="title"]'), '@title');
setHtml($$('label[for="volume"]'), '@volume');
setHtml($$('label[for="embed"]'), '@embed');
setHtml($$('label[for="autoplay"]'), '@autoplay');
setHtml($$('label[for="chat"]'), '@chat');
setHtml($$('label[for="force-fullscreen"]'), '@force_fullscreen');
setHtml($$('label[for="close-tab"]'), '@close_tab');
setHtml($$('label[for="no-cookie"]'), '@no_cookie');
setHtml($$('label[for="captions"]'), '@captions');
setHtml($$('label[for="annotations"]'), '@annotations');
setHtml($$('label[for="related"]'), '@related');
setHtml($$('label[for="controls"]'), '@controls');
setHtml($$('label[for="show-info"]'), '@show_info');
setHtml($$('label[for="fullscreen-button"]'), '@fullscreen_button');
setHtml($$('label[for="yt-logo"]'), '@yt_logo');
setHtml($$('label[for="keyboard"]'), '@keyboard');
setHtml($$('label[for="noclick"]'), '@noclick');
setHtml($$('label[for="loop"]'), '@loop');
setHtml($$('label[for="proportion"]'), '@proportion');
setHtml($$('label[for="youtube-api"]'), '@youtube_api');
setHtml($$('label[for="animate-title"]'), '@animate_title');
setHtml($$('label[for="shuffle"]'), '@shuffle');
setHtml($$('label[for="pause"]'), '@pause');
setHtml($$('label[for="spoof-referrer"]'), '@spoof_referrer');
setHtml($$('label[for="youtube-tv-on-error"]'), '@youtube_tv_on_error');
setHtml($$('label[for="force-youtube-tv"]'), '@force_youtube_tv');
setHtml($$('label[for="fix-popup"]'), '@fix_popup');
setHtml($$('label[for="app"]'), '@app');
setHtml($$('label[for="borderless"]'), '@borderless');
setHtml($$('label[for="always-on-top"]'), '@always_on_top');
setHtml($$('label[for="helium"]'), '@helium');
setHtml($$('label[for="helium-pin-tab"]'), '@helium_pin_tab');
setHtml($$('label[for="helium-version"]'), '@helium_version');
setHtml($$('label[for="keep-popup"]'), '@keep_popup');
setHtml($$('label[for="keep-dimensions"]'), '@keep_dimensions');
setHtml($$('label[for="spoof-user-agent"]'), '@spoof_user_agent');
setHtml($$('label[for="use-context"]'), '@use_context');
setHtml($$('label[for="enable-history"]'), '@enable_history');

setHtml($('requires-api'), '@requires_api');

var $defaultConfig = $('default-config');
setHtml($defaultConfig, '@default');

var $sourceCode = $('source-code');
setHtml($sourceCode, '@source_code');

var $instructions = $('instructions');
setHtml($instructions, '@instructions');

var $seeHistory = $('see-history');
setHtml($seeHistory, '@history');
// End Translation strings



var $align = $('align');
$align.value = options.align;
onChange($align, function() {
    setOption('align', +this.value);
});

var $width = $('width');
var $height = $('height');

$width.value = options.width;
onInput($width, function() {
    setOption('width', +this.value);
    highlightResolution(this.value, $height.value);
});

$height.value = options.height;
onInput($height, function() {
    setOption('height', +this.value);
    highlightResolution($width.value, this.value);
});

highlightResolution($width.value, $height.value);

var $horizontalMargin = $('horizontal-margin');
$horizontalMargin.value = options.horizontalMargin;
onInput($horizontalMargin, function() {
    setOption('horizontalMargin', +this.value);
});

var $verticalMargin = $('vertical-margin');
$verticalMargin.value = options.verticalMargin;
onInput($verticalMargin, function() {
    setOption('verticalMargin', +this.value);
});

var $color = $('color');
$color.value = options.color;
onChange($color, function() {
    setOption('color', this.value);
});

var $speed = $('speed');
$speed.value = options.speed;
onChange($speed, function() {
    setOption('speed', +this.value);
});

var $quality = $('quality');
$quality.value = options.quality;
onChange($quality, function() {
    setOption('quality', this.value);
});

var $title = $('title');
$title.value = options.title;
onInput($title, function() {
    setOption('title', this.value);
});

var $volume = $('volume');
var $currentVolume = $('current-volume');
$volume.value = options.volume;
$currentVolume.innerText = options.volume;
onInput($volume, function() {
    setOption('volume', +this.value);
    $currentVolume.innerText = this.value;
});

var $embed = $('embed');
$embed.checked = options.embed;
onChange($embed, function() {
    setOption('embed', this.checked);
});

var $autoplay = $('autoplay');
$autoplay.checked = options.autoplay;
onChange($autoplay, function() {
    setOption('autoplay', this.checked);
});

var $chat = $('chat');
$chat.checked = options.chat;
onChange($chat, function() {
    setOption('chat', this.checked);
});

var $forceFullscreen = $('force-fullscreen');
$forceFullscreen.checked = options.forceFullscreen;
onChange($forceFullscreen, function() {
    setOption('forceFullscreen', this.checked);
});

var $closeTab = $('close-tab');
$closeTab.checked = options.closeTab;
onChange($closeTab, function() {
    setOption('closeTab', this.checked);
});

var $noCookie = $('no-cookie');
$noCookie.checked = options.noCookie;
onChange($noCookie, function() {

    var $this = this;
    var isChecked = $this.checked;
    var permission = {
        origins: ['*://www.youtube-nocookie.com/*']
    };

    if (isChecked) {
        chrome.permissions.request(permission, function(granted) {
            $this.checked = granted;
            setOption('noCookie', granted);
        });
    }
    else {
        chrome.permissions.remove(permission, function(removed) {
            setOption('noCookie', false);
        });
    }
});

var $captions = $('captions');
$captions.checked = options.captions;
onChange($captions, function() {
    setOption('captions', this.checked);
});

var $annotations = $('annotations');
$annotations.checked = options.annotations;
onChange($annotations, function() {
    setOption('annotations', this.checked);
});

var $related = $('related');
$related.checked = options.related;
onChange($related, function() {
    setOption('related', this.checked);
});

var $controls = $('controls');
$controls.checked = options.controls;
onChange($controls, function() {
    setOption('controls', this.checked);
});

var $showInfo = $('show-info');
$showInfo.checked = options.showInfo;
onChange($showInfo, function() {
    setOption('showInfo', this.checked);
});

var $fullscreenButton = $('fullscreen-button');
$fullscreenButton.checked = options.fullscreenButton;
onChange($fullscreenButton, function() {
    setOption('fullscreenButton', this.checked);
});

var $ytLogo = $('yt-logo');
$ytLogo.checked = options.ytLogo;
onChange($ytLogo, function() {
    setOption('ytLogo', this.checked);
});

var $keyboard = $('keyboard');
$keyboard.checked = options.keyboard;
onChange($keyboard, function() {
    setOption('keyboard', this.checked);
});

var $noclick = $('noclick');
$noclick.checked = options.noclick;
onChange($noclick, function() {
    setOption('noclick', this.checked);
});

var $loop = $('loop');
$loop.checked = options.loop;
onChange($loop, function() {
    setOption('loop', this.checked);
});

var $proportion = $('proportion');
$proportion.checked = options.proportion;
onChange($proportion, function() {
    setOption('proportion', this.checked);
});

var $youtubeApi = $('youtube-api');
$youtubeApi.checked = options.youtubeApi;
onChange($youtubeApi, function() {
    setOption('youtubeApi', this.checked);
});

var $animateTitle = $('animate-title');
$animateTitle.checked = options.animateTitle;
onChange($animateTitle, function() {
    setOption('animateTitle', this.checked);
});

var $shuffle = $('shuffle');
$shuffle.checked = options.shuffle;
onChange($shuffle, function() {
    setOption('shuffle', this.checked);
});

var $pause = $('pause');
$pause.checked = options.pause;
onChange($pause, function() {
    setOption('pause', this.checked);
});

var $spoofReferrer = $('spoof-referrer');
$spoofReferrer.checked = options.spoofReferrer;
onChange($spoofReferrer, function() {
    setOption('spoofReferrer', this.checked);
    reloadBackground();
});

var $youtubeTvOnError = $('youtube-tv-on-error');
$youtubeTvOnError.checked = options.youtubeTvOnError;
onChange($youtubeTvOnError, function() {
    setOption('youtubeTvOnError', this.checked);
});

var $forceYoutubeTv = $('force-youtube-tv');
$forceYoutubeTv.checked = options.forceYoutubeTv;
onChange($forceYoutubeTv, function() {
    setOption('forceYoutubeTv', this.checked);
});

var $fixPopup = $('fix-popup');
$fixPopup.checked = options.fixPopup;
onChange($fixPopup, function() {
    setOption('fixPopup', this.checked);
});

var $app = $('app');
$app.checked = options.app;
onChange($app, function() {
    setOption('app', this.checked);
});

var $borderless = $('borderless');
$borderless.checked = options.borderless;
onChange($borderless, function() {
    setOption('borderless', this.checked);
});

var $alwaysOnTop = $('always-on-top');
$alwaysOnTop.checked = options.alwaysOnTop;
onChange($alwaysOnTop, function() {
    setOption('alwaysOnTop', this.checked);
});

var $helium = $('helium');
$helium.checked = options.helium;
onChange($helium, function() {
    setOption('helium', this.checked);
});

var $heliumPinTab = $('helium-pin-tab');
$heliumPinTab.checked = options.heliumPinTab;
onChange($heliumPinTab, function() {
    setOption('heliumPinTab', this.checked);
});

var $heliumVersion = $('helium-version');
$heliumVersion.value = options.heliumVersion;
onChange($heliumVersion, function() {
    setOption('heliumVersion', this.value);
});

var $keepPopup = $('keep-popup');
$keepPopup.checked = options.keepPopup;
onChange($keepPopup, function() {
    setOption('keepPopup', this.checked);
});

var $keepDimensions = $('keep-dimensions');
$keepDimensions.checked = options.keepDimensions;
onChange($keepDimensions, function() {
    setOption('keepDimensions', this.checked);
});

var $spoofUserAgent = $('spoof-user-agent');
$spoofUserAgent.checked = options.spoofUserAgent;
onChange($spoofUserAgent, function() {

    var $this = this;
    var isChecked = $this.checked;
    var permission = {
        origins: ['<all_urls>']
    };

    if (isChecked) {
        chrome.permissions.request(permission, function(granted) {
            $this.checked = granted;
            setOption('spoofUserAgent', granted);
            reloadBackground();
        });
    }
    else {
        setOption('spoofUserAgent', false);
        reloadBackground();
    }
});

var $context = $('use-context');
$context.checked = options.context;
onChange($context, function() {
    setOption('context', this.checked);
});

var $enableHistory = $('enable-history');
$enableHistory.checked = options.history;
onChange($enableHistory, function() {
    setOption('history', this.checked);
});

onClick($defaultConfig, function(e) {
    e.preventDefault();

    var isOption;

    if (confirm(getText('u_sure'))) {
        for (var i in localStorage) {
            if (localStorage.hasOwnProperty(i)) {
                isOption = i.substr(0, OPTIONS_PREFIX.length) === OPTIONS_PREFIX;

                if (isOption) {
                    localStorage.removeItem(i);
                }
            }
        }
        location.reload();
    }
});

onClick($sourceCode, function(e) {
    e.preventDefault();
    chrome.tabs.create({
        url: 'https://github.com/gabrielbarros/floating-player'
    });
});

onClick($instructions, function(e) {
    e.preventDefault();
    showInstructions();
});

onClick($seeHistory, function(e) {
    e.preventDefault();
    chrome.tabs.create({
        url: getExtensionUrl('history.html')
    });
});

onClick($('resolutions'), function(e) {
    var $target = e.target;
    var resolution;
    var width;
    var height;

    e.preventDefault();

    if ($target.href) {
        resolution = $target.innerText.split('x');
        width = resolution[0];
        height = resolution[1];

        $width.value = width;
        $height.value = height;

        setOption('width', +width);
        setOption('height', +height);

        highlightResolution(width, height);
    }
});

/*
Algorithm to generate "perfect" resolutions:

var width16x9;
var height16x9;
var width4x3;
var height4x3;

for (var i = 50; i < 500; i++) {
    height16x9 = height4x3 = i;
    width16x9 = (height16x9 * 16) / 9;
    width4x3 = (height4x3 * 4) / 3;

    if (width16x9 % 1 === 0 && width4x3 % 1 === 0) {
        console.log(width16x9 + 'x' + height16x9 +
            '; ' + width4x3 + 'x' + height4x3);
    }
}
*/

function highlightResolution(width, height) {
    var highlightClass = 'active-resolution';
    var $elem;

    $elem = $$('.' + highlightClass);
    $elem && $elem.classList.remove(highlightClass);

    $elem = $(width + 'x' + height);
    $elem && $elem.classList.add(highlightClass);
}


// Secret option to enable the add-on (https://git.io/vALr4) and set
// the popup always on top and make it borderless
addEvent(window, 'keyup', function(e) {
    var F9_KEY = 120;

    if (e.keyCode === F9_KEY) {

        chrome.permissions.request({
            permissions: ['management']
        },
        function(granted) {

            if (granted) {

                var addonId = localStorage.getItem('appid');
                var addonName = 'Addon for Floating Player (Chrome OS only)';

                // If appid is already set, then disable the extension
                if (addonId) {
                    chrome.management.setEnabled(addonId, false, function() {

                        // Extension not found
                        if (chrome.runtime.lastError) {
                            console.log(chrome.runtime.lastError.message);
                        }

                        // Remove permission
                        chrome.permissions.remove({
                            permissions: ['management']
                        });

                        localStorage.removeItem('appid');
                        setOption('app', false);
                        setOption('verticalMargin', 0);

                        alert(getText('addon_disabled'));
                    });
                }


                // If appid is not set, find the extension and enable it
                else {
                    chrome.management.getAll(function(result) {

                        // Loop through all installed extensions
                        for (var i = 0, len = result.length; i < len; i++) {
                            var extensionName = result[i].name;
                            var extensionId = result[i].id;

                            if (extensionName === addonName) {
                                addonId = extensionId;
                                break;
                            }
                        }

                        // Addon is not installed
                        if (addonId === null) {
                            alert(getText('addon_not_installed'));
                        }

                        // Addon is installed
                        else {
                            // Save addon id
                            localStorage.setItem('appid', addonId);

                            // Enable useful options
                            setOption('app', true);
                            setOption('borderless', true);
                            setOption('alwaysOnTop', true);

                            // Put the popup above the taskbar
                            if (options.align === ALIGN.BOTTOM_LEFT ||
                                options.align === ALIGN.BOTTOM_RIGHT ||
                                options.align === ALIGN.BOTTOM_CENTER) {

                                switch (userOs) {
                                    case OS.WINDOWS_XP:
                                    case OS.WINDOWS_VISTA:
                                        setOption('verticalMargin', 30);
                                        break;

                                    case OS.WINDOWS_7:
                                    case OS.WINDOWS_8:
                                    case OS.WINDOWS_8_1:
                                    case OS.WINDOWS_10:
                                        setOption('verticalMargin', 40);
                                        break;
                                }
                            }

                            // Enable extension if it's disabled
                            chrome.management.setEnabled(addonId, true, function() {

                                // Remove permission no longer needed
                                chrome.permissions.remove({
                                    permissions: ['management']
                                });

                                alert(getText('addon_enabled'));
                            });
                        }
                    });
                }
            }
            else {
                alert(getText('permission_not_granted'));
            }
        });
    }
});
